### ISSUE TITLE
[MASSEMBLY-675] Maven Assembly packaging wildcard-excluded dependencies

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#834]
**[Frank Wilson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zenlambda)** opened **[MASSEMBLY-675](https://issues.apache.org/jira/browse/MASSEMBLY-675?redirect=false)** and commented

Version 2.4 ignores wildcard exclusions in POM dependencies

Example (perhaps contrived - but easy to setup):

When a pom declares a dependency such as closure-compiler and for some reason we do not want to pull its dependencies in we could declare this in our POM, without having to know what those dependencies are:

```xml
<dependencies>
 <dependency>
 <groupId>com.google.javascript</groupId>
 <artifactId>closure-compiler</artifactId>
 <version>v20131014</version>
 <exclusions>
 <exclusion>
 <artifactId>*</artifactId>
 <groupId>*</groupId>
 </exclusion>
 </exclusions>
 </dependency>
 </dependencies>
```

This is a valid use of the exclusion feature as per [Maven Confluence](http://docs.codehaus.org/display/MAVENUSER/wildcard+exclusion+for+artifact+dependencies), MNG-3832. False warning about wildcards were [fixed](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commitdiff;h=65c135d5) in Git about 10 days ago
Here is the assembly descriptor:

```xml
<assembly>
 <id>bin</id>
 <formats>
 <format>dir</format>
 </formats>
 <includeBaseDirectory>false</includeBaseDirectory>
 <dependencySets>
 <dependencySet/>
 </dependencySets>
 </assembly>
```

We expect to only find the current project artifact and the closure-compiler JAR in our directory assembly. However the assembly plugin ignores our POM directive and includes the closure-compilers dependencies anyway!

Steps to reproduce are:

```
$ unzip massembly-675.zip
 $ cd massembly-675
 $ mvn clean install
 $ ls target/massembly-675-1-bin
 args4j-2.0.16.jar json-20090211.jar protobuf-java-2.4.1.jar
 closure-compiler-v20131014.jar jsr305-1.3.9.jar
 guava-15.0.jar massembly-675-1.jar
```

**Notice that the excluded jars are included in the assembly**

I would expect to only see the following JARs.
* closure-compiler-v20131014.jar
* massembly-675-1.jar

---

**Affects:** 2.4

**Attachments:**
- [massembly-675.zip](https://issues.apache.org/jira/secure/attachment/12714389/massembly-675.zip) (_2.37 kB_)

**Issue Links:**
- [MASSEMBLY-861](https://issues.apache.org/jira/browse/MASSEMBLY-861) exclusion * also packaged
 (_**"is duplicated by"**_)
- [MASSEMBLY-762](https://issues.apache.org/jira/browse/MASSEMBLY-762) Assembly plugin doesn't exclude transitive dependencies when excluded by wildcards in dependencies section

- [MASSEMBLY-799](https://issues.apache.org/jira/browse/MASSEMBLY-799) Exclusion on wildcard, then the assembly would still package to include the excluded libraries

**Remote Links:**
- [GitHub Pull Request #2
](https://github.com/apache/maven-assembly-plugin/pull/2)

19 votes, 21 watchers

**Comment by jira-importer** [comment]
**[Frank Wilson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zenlambda)** commented

Issue raised in error. (Or at lease I have to go back and look at my test case again!)

**Comment by jira-importer** [comment]
**[Frank Wilson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zenlambda)** commented

(Revised) Test case.

**Comment by jira-importer** [comment]
**[Frank Wilson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zenlambda)** commented

Reopening issue after revising description and test case.

The issue is with wildcard exclusions and not all exclusions as previously described.

**Comment by jira-importer** [comment]
**[bflorat](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bflorat)** commented

I'm really stuck with this issue. I have to build batchs as a archive with 178 transitive dependencies instead of ~30 when the exclusion wildcards is honored. some dependencies have to be excluded in order to make the program work so I have to explicitely exclude tens of dependencies instead of using the wildcards so these dependencies are not included into our archive.

Aside note : the dependency:tree plugin 2.1 had the same bug (presence of * excluded dependencies in the dep tree) but it has been fixed (in 2.8 or before).

Thanks

**Comment by jira-importer** [comment]
**[Tomer Cohen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tomerc)** commented

+1 for this issue

**Comment by jira-importer** [comment]
**[Matthias Müller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matthias_82)** commented

Is there any chance to get this fixed? From the unresolved tickets in this project, it's the one with the most votes...

**Comment by jira-importer** [comment]
**[Rodrigo Carvalho Silva](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rcsilva83)** commented

I lost one day of work trying to figure out why my package was being created with undesired dependencies...

Is this issue difficult to be fixed by someone with no experience with the the plugin's codebase (like me)?

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

Fixed with initial PR in [b0cf85db9c466d42a4d39f46634ad49cbb784d8c](https://gitbox.apache.org/repos/asf?p=maven-assembly-plugin.git;a=commit;h=b0cf85db9c466d42a4d39f46634ad49cbb784d8c) and additional [78faddf88b1537bc5c61ba74cdb2b227416152b6](https://gitbox.apache.org/repos/asf?p=maven-assembly-plugin.git;a=commit;h=78faddf88b1537bc5c61ba74cdb2b227416152b6).
